mirgecom/fluid.py

```diff
@@ -0,0 +1,62 @@
+""":mod:`mirgecom.fluid` provides common utilities for fluid simulation.
+
+.. autofunction:: compute_velocity_gradient
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+import numpy as np  # noqa
+from meshmode.dof_array import DOFArray  # noqa
+from mirgecom.euler import ConservedVars
+
+
+def compute_velocity_gradient(discr, cv: ConservedVars):
+    r"""
+    Compute the gradient of fluid velocity.
+
+    Computes the gradient of fluid velocity from:
+
+    .. math::
+
+        \nabla{v_i} = \frac{1}{\rho}(\nabla(\rho{v_i})-v_i\nabla{\rho}),
+
+    where $v_i$ is ith velocity component.
+
+    Parameters
+    ----------
+    discr: grudge.eager.EagerDGDiscretization
+        the discretization to use
+    cv: mirgecom.euler.ConservedVars
+        the fluid conserved variables
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray`
+        representing $\partial_j{v_i}$.
+    """
+    dim = discr.dim
+    velocity = cv.momentum/cv.mass
+    dmass = discr.grad(cv.mass)
+    dmom = [discr.grad(cv.momentum[i]) for i in range(dim)]
+    return [(dmom[i] - velocity[i]*dmass)/cv.mass for i in range(dim)]
```

Comment:
My one concern with this is that, in the presence of solutions with discontinuities (as you're likely to find with DG), this is not a consistent gradient. Some type of flux term is needed.

Reply:
Sure. Is there an existing `grad` utility that can help me with that, or do I need to add an interface-aware grad?  Is there a grudge-oriented example I could look at?

Reply:
To "just move past this", a central flux ("just average both sides") would at least be consistent, but IMO not desirable. Upwinding is a good form of dissipation to have, after all. :) So there's a whole design space here that depends on how you'd like to discretize the overall operator, and the choice may be different for viscous and inviscid terms. (but do note that the volume derivatives can be reused) For perspective, note how local Lax-Friedrichs is just central + a penalty. For second order operators, "unstabilized central", "interior penalty", and "LDG" are the options that are worth considering IMO.

This is DG as a scheme being "not nice" to the code-driven desire to build abstractions (like the one you're proposing here).

Summoning @thomasgibson, maybe he has additional perspective.

cc @lukeolson

Reply:
Just so we're clear, this part is supposed to be only the volume piece evaluated at the nodes of the discretization.  The fluxes I plan on constructing with a symmetric tensor built from this function.

Reply:
So just so I understand, is this function just returning the cell-wise gradient of velocity? Meaning that it's not really a "DG gradient"? In which case, this should be made clear in the docstring that this is basically computing `\nabla|_K u`, where `K` is some cell. There are no integrals here?